File: aoc2019/day5.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.IO;
using System.Linq;

namespace aoc2019
{
    public class day5
    {
        public static void Go()
        {
            string input = File.ReadLines(@"c:\github\aoc2019\input\day5.txt").First();
            List<int> items = input.Split(',').Select(Int32.Parse).ToList();
            Console.WriteLine("PartA:");
            run(items, 1);
            Console.WriteLine("PartB:");
            items = input.Split(',').Select(Int32.Parse).ToList();
            run(items, 5);
        }

        private static int getparam(List<int> items, int paramnum, int pointer, bool position)
        {
            if (position)
                return items[items[pointer + paramnum]];
            else
                return items[pointer + paramnum];
        }

        private static int run(List<int> items, int machineocode)
        {
            int pointer = 0;
            List<int> param1needed = new List<int> { 1, 2, 4, 5, 6, 7, 8 };
            List<int> param2needed = new List<int> { 1, 2, 5, 6, 7, 8 };

            while (true)
            {
                string opcodes = items[pointer].ToString();

                while (opcodes.Length < 5)
                    opcodes = "0" + opcodes;

                int mode1 = int.Parse(opcodes.Substring(2, 1));
                int mode2 = int.Parse(opcodes.Substring(1, 1));
                int mode3 = int.Parse(opcodes.Substring(0, 1));
                int opcode = int.Parse(opcodes.Substring(3, 2));

                int param1=0;
                int param2=0;

                if (param1needed.Contains(opcode))
                    param1 = getparam(items, 1, pointer, mode1 == 0);
                if (param2needed.Contains(opcode))
                    param2 = getparam(items, 2, pointer, mode2 == 0);

                switch (opcode)
                {
                    case 99:
                        Console.WriteLine("STOP");
                        return 0;
                    case 1:
                        items[items[pointer + 3]] = param1 + param2;
                        pointer += 4;
                        break;
                    case 2:
                        items[items[pointer + 3]] = param1 * param2;
                        pointer += 4;
                        break;
                    case 3:
                        items[items[pointer + 1]] = machineocode;
                        pointer += 2;
                        break;
                    case 4:
                        Console.WriteLine(param1);
                        pointer += 2;
                        break;
                    case 5:
                        if (param1 != 0)
                            pointer = param2;
                        else
                            pointer += 3;
                        break;
                    case 6:
                        if (param1 == 0)
                            pointer = param2;
                        else
                            pointer += 3;
                        break;
                    case 7:
                        if (param1 < param2)
                            items[items[pointer + 3]] = 1;
                        else
                            items[items[pointer + 3]] = 0;
                        pointer += 4;
                        break;
                    case 8:
                        if (param1 == param2)
                            items[items[pointer + 3]] = 1;
                        else
                            items[items[pointer + 3]] = 0;
                        pointer += 4;
                        break;
                    default:
                        break;
                }
            }
        }
    }
}

File: python/day5.py
def run(nums,input):
    pointer=0
    while True:
        t=f"{nums[pointer]:05}"
        opcode,a,b,c=int(t[3:]),int(t[0]),int(t[1]),int(t[2])
        if opcode in[ 1, 2, 4, 5, 6, 7, 8]:
            par1 = nums[nums[pointer + 1]] if c==0 else nums[pointer + 1]
        if opcode in[ 1, 2, 5, 6, 7, 8]:
            par2 = nums[nums[pointer + 2]] if b==0 else nums[pointer + 2]
        if opcode == 99:
            break
        elif opcode == 1:
            nums[nums[pointer + 3]] = par1 + par2
        elif opcode == 2:
            nums[nums[pointer + 3]] = par1 * par2
        elif opcode == 3:
            nums[nums[pointer + 1]] = input
        elif opcode == 4:
            if par1!=0: print(par1)
        elif opcode == 5:
            if par1 !=0: pointer = par2 
            else: pointer += 3
        elif opcode == 6:
            if par1 ==0: pointer = par2 
            else: pointer += 3
        elif opcode == 7:
            if par1 < par2: nums[nums[pointer + 3]] = 1 
            else: nums[nums[pointer + 3]] = 0
        elif opcode == 8:
            if par1 == par2 : nums[nums[pointer + 3]] = 1 
            else: nums[nums[pointer + 3]] = 0
        if opcode in [1, 2, 7, 8]:
            pointer+=4
        if opcode in [3,4]:
            pointer+=2

f = open("..\input\day5.txt", "r")
input = f.read()
f.close()
nums = [int(n) for n in input.split(',')]
print ("AnswerA: ",end='') 
run(nums.copy(),1)
print ("AnswerB: ",end='')
run(nums,5)